README.md
# pick-alternate-value

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Coveralls Status][coveralls-image]][coveralls-url]
[![Dependency Status][depstat-image]][depstat-url]
[![Downloads][download-badge]][npm-url]

> Pick the most suitable value from a list of possible values

## Install

```sh
npm i -D pick-alternate-value
```

## Usage

```js
import pickAlternateValue from "pick-alternate-value"

pickAlternateValue() // true
```

## Functions

<dl>
<dt><a href="#pickLongestSize">pickLongestSize(list, defaultValue, max)</a> ⇒ <code>object</code></dt>
<dd><p>Pick the first string (or object) with the longest size</p>
</dd>
<dt><a href="#pickShortestSize">pickShortestSize(list, defaultValue, min)</a> ⇒ <code>object</code></dt>
<dd><p>Pick the first string (or object) with the shortest size</p>
</dd>
<dt><a href="#sumSize">sumSize(list)</a> ⇒ <code>integer</code></dt>
<dd><p>Adds the size of all the items in the list</p>
</dd>
<dt><a href="#minSize">minSize(list)</a> ⇒ <code>integer</code></dt>
<dd><p>Finds the minimum size of all the items in the list</p>
</dd>
<dt><a href="#maxSize">maxSize(list)</a> ⇒ <code>integer</code></dt>
<dd><p>Finds the maximum size of all the items in the list</p>
</dd>
<dt><a href="#hasNoNull">hasNoNull(list)</a> ⇒ <code>integer</code></dt>
<dd><p>Returns true if the list does not contain any null</p>
</dd>
<dt><a href="#discardPlaceholders">discardPlaceholders(template, phStart, phEnd)</a> ⇒ <code>string</code></dt>
<dd><p>Discard the placeholders of a string template. Useful to check
the minimum length of a template.</p>
</dd>
<dt><a href="#getArrInArr">getArrInArr(arrIdx, listOfList)</a> ⇒ <code>array</code></dt>
<dd><p>Gets an array using an array of indexes</p>
</dd>
<dt><a href="#decArrayIndex">decArrayIndex(arrIdx, maxIdx)</a> ⇒ <code>array</code></dt>
<dd><p>Decrements an array of indexes from left to right.
For right to left, uses .reverse().
This is useful for iterating over indices.</p>
</dd>
<dt><a href="#combineListOfList">combineListOfList(listOfList)</a> ⇒ <code>array</code></dt>
<dd><p>Combine a list of list in a similar as imbricated for loops</p>
</dd>
<dt><a href="#highestRankedCombination">highestRankedCombination(listCombination, rankFn, filterFn)</a> ⇒ <code>array</code></dt>
<dd><p>Finds the combination with the highest rank</p>
</dd>
</dl>

<a name="pickLongestSize"></a>

## pickLongestSize(list, defaultValue, max) ⇒ <code>object</code>
Pick the first string (or object) with the longest size

**Kind**: global function  
**Returns**: <code>object</code> - The object or string that is the longest  

| Param | Type | Description |
| --- | --- | --- |
| list | <code>array</code> | list of strings or objects |
| defaultValue | <code>object</code> | the object/string to return if null |
| max | <code>integer</code> | the maximum size that is allowed |

**Example**  
```js
// returns abcd
pav.pickLongestSize(['ab', 'a', 'abcd'])
```
<a name="pickShortestSize"></a>

## pickShortestSize(list, defaultValue, min) ⇒ <code>object</code>
Pick the first string (or object) with the shortest size

**Kind**: global function  
**Returns**: <code>object</code> - The object or string that is the longest  

| Param | Type | Description |
| --- | --- | --- |
| list | <code>array</code> | list of strings or objects |
| defaultValue | <code>object</code> | the object/string to return if null |
| min | <code>integer</code> | the minimum size that is allowed |

**Example**  
```js
// returns a1
pav.pickShortestSize(['abd', 'a1', 'a2', 'abc'])
```
<a name="sumSize"></a>

## sumSize(list) ⇒ <code>integer</code>
Adds the size of all the items in the list

**Kind**: global function  
**Returns**: <code>integer</code> - The sum of all the sizes  

| Param | Type | Description |
| --- | --- | --- |
| list | <code>array</code> | list of strings or objects |

**Example**  
```js
// returns 7
pav.sumSize(['ab', 'a', 'abcd'])
```
<a name="minSize"></a>

## minSize(list) ⇒ <code>integer</code>
Finds the minimum size of all the items in the list

**Kind**: global function  
**Returns**: <code>integer</code> - The minimum of all the sizes  

| Param | Type | Description |
| --- | --- | --- |
| list | <code>array</code> | list of strings or objects |

**Example**  
```js
// returns 1
pav.minSize(['ab', 'a', 'abcd'])
```
<a name="maxSize"></a>

## maxSize(list) ⇒ <code>integer</code>
Finds the maximum size of all the items in the list

**Kind**: global function  
**Returns**: <code>integer</code> - The maximum of all the sizes  

| Param | Type | Description |
| --- | --- | --- |
| list | <code>array</code> | list of strings or objects |

**Example**  
```js
// returns 4
pav.maxSize(['ab', 'a', 'abcd'])
```
<a name="hasNoNull"></a>

## hasNoNull(list) ⇒ <code>integer</code>
Returns true if the list does not contain any null

**Kind**: global function  
**Returns**: <code>integer</code> - true if no null  

| Param | Type | Description |
| --- | --- | --- |
| list | <code>array</code> | list of strings or objects |

**Example**  
```js
// returns false
pav.hasNoNull(['ab', null, 'abcd'])
```
<a name="discardPlaceholders"></a>

## discardPlaceholders(template, phStart, phEnd) ⇒ <code>string</code>
Discard the placeholders of a string template. Useful to check
the minimum length of a template.

**Kind**: global function  
**Returns**: <code>string</code> - The template without ny placeholders  

| Param | Type | Description |
| --- | --- | --- |
| template | <code>string</code> | list of strings or objects |
| phStart | <code>string</code> | the placeholder start keyword |
| phEnd | <code>string</code> | the placeholder end keyword |

**Example**  
```js
// returns 123456
pav.discardPlaceholders('1234${placeholder}56','${','}')
```
<a name="getArrInArr"></a>

## getArrInArr(arrIdx, listOfList) ⇒ <code>array</code>
Gets an array using an array of indexes

**Kind**: global function  
**Returns**: <code>array</code> - the selected list  

| Param | Type | Description |
| --- | --- | --- |
| arrIdx | <code>array</code> | an array of indices |
| listOfList | <code>array</code> | a list of list |

**Example**  
```js
// returns ['b', '1']
pav.getArrInArr([1, 1], [['a','b'], [1, 2]])
```
<a name="decArrayIndex"></a>

## decArrayIndex(arrIdx, maxIdx) ⇒ <code>array</code>
Decrements an array of indexes from left to right.
For right to left, uses .reverse().
This is useful for iterating over indices.

**Kind**: global function  
**Returns**: <code>array</code> - the selected list  

| Param | Type | Description |
| --- | --- | --- |
| arrIdx | <code>array</code> | an array of indices |
| maxIdx | <code>array</code> | an array of the maximum indices |

**Example**  
```js
// returns [1, 2, 4]
pav.decArrayIndex([2, 2, 4], [3, 2, 4])
```
<a name="combineListOfList"></a>

## combineListOfList(listOfList) ⇒ <code>array</code>
Combine a list of list in a similar as imbricated for loops

**Kind**: global function  
**Returns**: <code>array</code> - All the possible combinations in reverse order  

| Param | Type | Description |
| --- | --- | --- |
| listOfList | <code>array</code> | a list of list |

**Example**  
```js
// returns [['b', 2], ['a', 2], ['b', 1], ['a', 1]]
pav.combineListOfList([['a','b'], [1, 2]])
```
<a name="highestRankedCombination"></a>

## highestRankedCombination(listCombination, rankFn, filterFn) ⇒ <code>array</code>
Finds the combination with the highest rank

**Kind**: global function  
**Returns**: <code>array</code> - The highest ranked combination  

| Param | Type | Description |
| --- | --- | --- |
| listCombination | <code>array</code> | a list of list |
| rankFn | <code>function</code> | a function which returns the rank. Default by size |
| filterFn | <code>function</code> | a function which filter only the suitable combination. |

**Example**  
```js
// returns ['bb', 22]
pav.highestRankedCombination([['bb', 2], ['a', 22], ['b', 1], ['a', 1]])
```


## License

MIT © [olih](http://github.com/flarebyte)

[npm-url]: https://npmjs.org/package/pick-alternate-value
[npm-image]: https://img.shields.io/npm/v/pick-alternate-value.svg?style=flat-square

[travis-url]: https://travis-ci.org/flarebyte/pick-alternate-value
[travis-image]: https://img.shields.io/travis/flarebyte/pick-alternate-value.svg?style=flat-square

[coveralls-url]: https://coveralls.io/r/flarebyte/pick-alternate-value
[coveralls-image]: https://img.shields.io/coveralls/flarebyte/pick-alternate-value.svg?style=flat-square

[depstat-url]: https://david-dm.org/flarebyte/pick-alternate-value
[depstat-image]: https://david-dm.org/flarebyte/pick-alternate-value.svg?style=flat-square

[download-badge]: http://img.shields.io/npm/dm/pick-alternate-value.svg?style=flat-square


CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
# Changelog

- [v1.0.0](#v100)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### v1.0.0

 * pick-alternate-value initial commit.


test/index.js
import test from 'tape';
import pav from '../src';


const a = 'A';
const b = 'B';
const c = 'C';
const a31 = ['aaa', 'a'];
const a123 = ['b', 'bb', 'bbb'];
const a2135 = ['cc', 'c', 'ccc', 'ccccc'];


test('pick the longest size', (t) => {
  t.plan(10);
  t.equal(pav.pickLongestSize(null), null, 'null');
  t.equal(pav.pickLongestSize([]), null, 'empty');
  t.equal(pav.pickLongestSize(null, 'def123'), 'def123', 'null');
  t.equal(pav.pickLongestSize([null, '']), '', 'discard null');
  t.equal(pav.pickLongestSize(['ab', 'a', 'abcd'], null, 2), 'ab', 'with max');
  t.equal(pav.pickLongestSize(['ab', 'a', 'abcd']), 'abcd', 'A');
  t.equal(pav.pickLongestSize(['abcde', 'a', 'ab']), 'abcde', 'B');
  t.equal(pav.pickLongestSize(['ab', 'ab1', 'ab2', 'a']), 'ab1', 'take 1st of 2 eq');
  t.deepEqual(pav.pickLongestSize([[1, 2], [1, 2, 3], [1]]), [1, 2, 3], 'arr of array');
  t.deepEqual(pav.pickLongestSize([{ a, b }, { a, b, c }, { a }]), { a, b, c }, 'array of obj');
});

test('pick the shortest size', (t) => {
  t.plan(10);
  t.equal(pav.pickShortestSize(null), null, 'null');
  t.equal(pav.pickShortestSize([]), null, 'empty');
  t.equal(pav.pickShortestSize([], 'def123'), 'def123', 'empty');
  t.equal(pav.pickShortestSize([null, '', 'abc']), '', 'discard null');
  t.equal(pav.pickShortestSize(['ab', 'a', 'abcd']), 'a', 'A');
  t.equal(pav.pickShortestSize(['ab', 'a', 'abcd'], null, 2), 'ab', 'with min');
  t.equal(pav.pickShortestSize(['abcde', 'a', 'ab']), 'a', 'B');
  t.equal(pav.pickShortestSize(['abd', 'a1', 'a2', 'abc']), 'a1', 'take 1st of 2 eq');
  t.deepEqual(pav.pickShortestSize([[1, 2], [1, 2, 3], [1]]), [1], 'arr of array');
  t.deepEqual(pav.pickShortestSize([{ a, b }, { a, b, c }, { a }]), { a }, 'array of obj');
});


test('sum the size', (t) => {
  t.plan(6);
  t.equal(pav.sumSize(null), 0, 'zero');
  t.equal(pav.sumSize([]), 0, 'empty');
  t.equal(pav.sumSize(['ab', 'a', 'abcd']), 7, 'A');
  t.equal(pav.sumSize(['ab', 'a', null]), 3, 'with null');
  t.deepEqual(pav.sumSize([[1, 2], [1, 2, 3], [1]]), 6, 'arr of array');
  t.deepEqual(pav.sumSize([{ a, b }, { a, b, c }, { a }]), 6, 'array of obj');
});

test('max the size', (t) => {
  t.plan(6);
  t.equal(pav.maxSize(null), null, 'zero');
  t.equal(pav.maxSize([]), null, 'empty');
  t.equal(pav.maxSize(['ab', 'a', 'abcd']), 4, 'A');
  t.equal(pav.maxSize(['ab', 'a', null]), 2, 'with null');
  t.deepEqual(pav.maxSize([[1, 2], [1, 2, 3], [1]]), 3, 'arr of array');
  t.deepEqual(pav.maxSize([{ a, b }, { a, b, c }, { a }]), 3, 'array of obj');
});

test('min the size', (t) => {
  t.plan(6);
  t.equal(pav.minSize(null), null, 'zero');
  t.equal(pav.minSize([]), null, 'empty');
  t.equal(pav.minSize(['ab', 'a', 'abcd']), 1, 'A');
  t.equal(pav.minSize(['ab', 'a', null]), 1, 'with null');
  t.deepEqual(pav.minSize([[1, 2], [1, 2, 3], [1]]), 1, 'arr of array');
  t.deepEqual(pav.minSize([{ a, b }, { a, b, c }, { a }]), 1, 'array of obj');
});

test('has no null', (t) => {
  t.plan(2);
  t.equal(pav.hasNoNull([1, 'a', 'aa']), true, 'no null');
  t.equal(pav.hasNoNull([1, null, 'aa']), false, 'empty');
});

test('discard placeholders', (t) => {
  t.plan(2);
  const p = '${';
  const given = `12345${p} user }6${p}a.b1.c}7${p}sum(10,15)}`;
  t.equal(pav.discardPlaceholders(null), null, 'with null');
  t.equal(pav.discardPlaceholders(given), '1234567', '');
});

test('get an array inside an array', (t) => {
  t.plan(3);
  t.deepEqual(pav.getArrInArr([0, 0, 0], [a31, a123, a2135]),
   [a31[0], a123[0], a2135[0]], 'first');
  t.deepEqual(pav.getArrInArr([1, 1, 1], [a31, a123, a2135]),
    [a31[1], a123[1], a2135[1]], 'second');
  t.deepEqual(pav.getArrInArr([5, 7, 8], [a31, a123, a2135]),
     null, 'none');
});

test('decrement an index of array starting from head', (t) => {
  t.plan(12);
  t.deepEqual(pav.decArrayIndex([3, 2], [3, 2]), [2, 2], 'A');
  t.deepEqual(pav.decArrayIndex([0, 2], [3, 2]), [3, 1], 'B');
  t.deepEqual(pav.decArrayIndex([0, 1], [3, 2]), [3, 0], 'C');
  t.deepEqual(pav.decArrayIndex([1, 0], [3, 2]), [0, 0], 'D');
  t.deepEqual(pav.decArrayIndex([0, 0], [3, 2]), [-1, 0], 'E');
  t.deepEqual(pav.decArrayIndex([2, 2, 4], [3, 2, 4]), [1, 2, 4], 'F');
  t.deepEqual(pav.decArrayIndex([0, 2, 4], [3, 2, 4]), [3, 1, 4], 'G');
  t.deepEqual(pav.decArrayIndex([0, 0, 4], [3, 2, 4]), [3, 2, 3], 'H');
  t.deepEqual(pav.decArrayIndex([1, 0, 0], [3, 2, 4]), [0, 0, 0], 'I');
  t.deepEqual(pav.decArrayIndex([0, 0, 0], [3, 2, 4]), [-1, 0, 0], 'J');
  t.deepEqual(pav.decArrayIndex([0, 0, 0, 0], [3, 2, 4, 5]), [-1, 0, 0, 0], 'K');
  t.deepEqual(pav.decArrayIndex([0, 1, 0, 0], [3, 2, 4, 5]), [3, 0, 0, 0], 'L');
});


test('combine a list of list', (t) => {
  let i;
  let j;
  let k;
  t.plan(2);
  const expected = [];

  for (i of a2135) {
    for (j of a123) {
      for (k of a31) {
        expected.push([k, j, i]);
      }
    }
  }
  const actual = pav.combineListOfList([a31, a123, a2135]);
  t.equal(actual.length, 2 * 3 * 4, 'size');
  t.deepEqual(actual, expected.reverse(), 'Table');
});

test('get highest ranked combination', (t) => {
  let i;
  let j;
  let k;
  t.plan(1);
  const combi = [];

  for (i of a2135) {
    for (j of a123) {
      for (k of a31) {
        combi.push([k, j, i]);
      }
    }
  }

  const actual = pav.highestRankedCombination(combi);
  t.deepEqual(actual, ['aaa', 'bbb', 'ccccc'], 'maximum');
});

test('get highest ranked combination with null', (t) => {
  let i;
  let j;
  let k;
  t.plan(1);
  const combi = [];

  for (i of a2135) {
    for (j of ['b', 'bb', null]) {
      for (k of a31) {
        combi.push([k, j, i]);
      }
    }
  }

  const actual = pav.highestRankedCombination(combi);
  t.deepEqual(actual, ['aaa', 'bb', 'ccccc'], 'maximum');
});

test('get lowest ranked combination', (t) => {
  let i;
  let j;
  let k;
  t.plan(1);
  const combi = [];

  for (i of a2135) {
    for (j of ['b', 'bb', null]) {
      for (k of a31) {
        combi.push([k, j, i]);
      }
    }
  }

  const actual = pav.highestRankedCombination(combi, l => -1 * pav.sumSize(l));
  t.deepEqual(actual, ['a', 'b', 'c'], 'minimum');
});


src/index.js
import _ from 'lodash';

/**
 * Pick the first string (or object) with the longest size
 * @param {array} list - list of strings or objects
 * @param {object} defaultValue - the object/string to return if null
 * @param {integer} max - the maximum size that is allowed
 * @example
 * // returns abcd
 * pav.pickLongestSize(['ab', 'a', 'abcd'])
 * @return {object} The object or string that is the longest
 */
const pickLongestSize = (list,
   defaultValue = null,
   max = Number.MAX_SAFE_INTEGER) => {
  const noNullList = _.filter(list, s => !_.isNil(s) && _.size(s) <= max);
  if (_.isEmpty(noNullList)) return defaultValue;
  const sorted = _.sortBy(noNullList, s => _.size(s));
  const maxSize = _.size(_.last(sorted));
  const filtered = _.filter(noNullList, s => _.size(s) === maxSize);
  return _.head(filtered);
};

/**
 * Pick the first string (or object) with the shortest size
 * @param {array} list - list of strings or objects
 * @param {object} defaultValue - the object/string to return if null
 * @param {integer} min - the minimum size that is allowed
 * @example
 * // returns a1
 * pav.pickShortestSize(['abd', 'a1', 'a2', 'abc'])
 * @return {object} The object or string that is the longest
 */
const pickShortestSize = (list,
  defaultValue = null,
  min = -1
  ) => {
  const noNullList = _.filter(list, s => !_.isNil(s) && _.size(s) >= min);
  if (_.isEmpty(noNullList)) return defaultValue;
  const sorted = _.sortBy(noNullList, s => _.size(s));
  const minSize = _.size(_.head(sorted));
  const filtered = _.filter(noNullList, s => _.size(s) === minSize);
  return _.head(filtered);
};

/**
 * Adds the size of all the items in the list
 * @param {array} list - list of strings or objects
 * @example
 * // returns 7
 * pav.sumSize(['ab', 'a', 'abcd'])
 * @return {integer} The sum of all the sizes
 */
const sumSize = list =>
   _.sumBy(list, s => _.size(s))
;

/**
 * Finds the minimum size of all the items in the list
 * @param {array} list - list of strings or objects
 * @example
 * // returns 1
 * pav.minSize(['ab', 'a', 'abcd'])
 * @return {integer} The minimum of all the sizes
 */
const minSize = (list) => {
  const noNullList = _.filter(list, s => !_.isNil(s));
  if (_.isEmpty(noNullList)) return null;
  const sorted = _.sortBy(noNullList, s => _.size(s));
  return _.size(_.head(sorted));
};

/**
 * Finds the maximum size of all the items in the list
 * @param {array} list - list of strings or objects
 * @example
 * // returns 4
 * pav.maxSize(['ab', 'a', 'abcd'])
 * @return {integer} The maximum of all the sizes
 */
const maxSize = (list) => {
  const noNullList = _.filter(list, s => !_.isNil(s));
  if (_.isEmpty(noNullList)) return null;
  const sorted = _.sortBy(noNullList, s => _.size(s));
  return _.size(_.last(sorted));
};

/**
 * Returns true if the list does not contain any null
 * @param {array} list - list of strings or objects
 * @example
 * // returns false
 * pav.hasNoNull(['ab', null, 'abcd'])
 * @return {integer} true if no null
 */
const hasNoNull = list => !_.some(list, _.isNil);

/**
 * Discard the placeholders of a string template. Useful to check
 * the minimum length of a template.
 * @param {string} template - list of strings or objects
 * @param {string} phStart - the placeholder start keyword
 * @param {string} phEnd - the placeholder end keyword
 * @example
 * // returns 123456
 * pav.discardPlaceholders('1234${placeholder}56','${','}')
 * @return {string} The template without ny placeholders
 */
const discardPlaceholders = (template, phStart = '${', phEnd = '}') => {
  if (_.isNil(template)) return null;
  const anyNames = '[A-Za-z0-9(),.]+';
  const phS = _.escapeRegExp(phStart);
  const phE = _.escapeRegExp(phEnd);
  const search = new RegExp(`${phS}\\s*${anyNames}\\s*${phE}`, 'g');
  const withoutVars = template.replace(search, '');
  return withoutVars;
};

/**
 * Gets an array using an array of indexes
 * @param {array} arrIdx - an array of indices
 * @param {array} listOfList - a list of list
 * @example
 * // returns ['b', '1']
 * pav.getArrInArr([1, 1], [['a','b'], [1, 2]])
 * @return {array} the selected list
 */
const getArrInArr = (arrIdx, listOfList) => {
  if (_.isNil(arrIdx) || _.isNil(listOfList)) return null;
  const sizeIdx = _.size(arrIdx);
  const sizeList = _.size(listOfList);
  if (sizeIdx !== sizeList) {
    throw new Error(`size idx ${sizeIdx} different of ${sizeList}`);
  }
  const result = [];
  for (let i = 0; i < sizeIdx; i += 1) {
    const value = listOfList[i][arrIdx[i]];
    if (_.isUndefined(value)) {
      return null;
    }
    result.push(value);
  }
  return result;
};

/**
 * Decrements an array of indexes from left to right.
 * For right to left, uses .reverse().
 * This is useful for iterating over indices.
 * @param {array} arrIdx - an array of indices
 * @param {array} maxIdx - an array of the maximum indices
 * @example
 * // returns [1, 2, 4]
 * pav.decArrayIndex([2, 2, 4], [3, 2, 4])
 * @return {array} the selected list
 */
const decArrayIndex = (arrIdx, maxIdx) => {
  const arrIdxl = _.cloneDeep(arrIdx);
  const maxIdxl = _.cloneDeep(maxIdx);
  const headDec = arrIdxl[0] - 1;
  if (_.size(arrIdxl) === 1) {
    return [headDec];
  }
  const tailIdx = _.tail(arrIdxl);
  const tailMaxIdx = _.tail(maxIdxl);
  const isAllZerosOnRight = _.every(tailIdx, n => n === 0);
  if (headDec < 0 && !isAllZerosOnRight) {
    const tailRes = decArrayIndex(tailIdx, tailMaxIdx);
    tailRes.unshift(maxIdxl[0]);
    return tailRes;
  }
  tailIdx.unshift(headDec);
  return tailIdx;
};

/**
 * Combine a list of list in a similar as imbricated for loops
 * @param {array} listOfList - a list of list
 * @example
 * // returns [['b', 2], ['a', 2], ['b', 1], ['a', 1]]
 * pav.combineListOfList([['a','b'], [1, 2]])
 * @return {array} All the possible combinations in reverse order
 */
const combineListOfList = (listOfList) => {
  if (_.isEmpty(listOfList)) return null;
  const maxIdx = _.map(listOfList, n => _.size(n) - 1);
  let idx = maxIdx;
  const result = [];

  while (idx[0] !== -1) {
    const list = getArrInArr(idx, listOfList);
    result.push(list);
    idx = decArrayIndex(idx, maxIdx);
  }

  return result;
};

/**
 * Finds the combination with the highest rank
 * @param {array} listCombination - a list of list
 * @param {function} rankFn - a function which returns the rank. Default by size
 * @param {function} filterFn - a function which filter only the suitable
 * combination.
 * @example
 * // returns ['bb', 22]
 * pav.highestRankedCombination([['bb', 2], ['a', 22], ['b', 1], ['a', 1]])
 * @return {array} The highest ranked combination
 */
const highestRankedCombination = (listCombination,
  rankFn = sumSize, filterFn = hasNoNull) => {
  const filtered = _.filter(listCombination, filterFn);
  if (_.isEmpty(filtered)) return null;
  const sorted = _.sortBy(filtered, rankFn);
  return _.last(sorted);
};


const pickAlternateValue = {
  pickLongestSize,
  pickShortestSize,
  sumSize,
  minSize,
  maxSize,
  hasNoNull,
  discardPlaceholders,
  getArrInArr,
  decArrayIndex,
  combineListOfList,
  highestRankedCombination,
};

module.exports = pickAlternateValue;
